File: Linked_List/Singly Linked List/LinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        
class LinkedList:
    def __init__(self):
        self.head = None
        
    def insertAtLast(self, data):
        node = Node(data)
        if self.head == None:
            self.head = node
            return

        currPointer = self.head
        
        while currPointer.next != None:
            currPointer = currPointer.next
        currPointer.next = node
        
    def deleteNode(self, dataKey):
        prevNode = None
        currNode = self.head
        
        # if head element is to be deleted
        if currNode.data == dataKey:
            currNode = currNode.next
            self.head = currNode
            return
        
        while currNode:
            if currNode.data == dataKey:
                prevNode.next = currNode.next
                return
            prevNode = currNode
            currNode = currNode.next
        
        print('Node with data key not found')
        
    def __str__(self):
        result = ''
        currNode = self.head
        while currNode:
            if currNode.next != None:
                result += str(currNode.data) + ' -> '
            else:
                result += str(currNode.data)
            
            currNode = currNode.next
        
        return result
            
            

if __name__ == "__main__":
    linkedList = LinkedList()
    
    for i in range(10):
        linkedList.insertAtLast(i)
        
    print(linkedList)
    
    # Deletion from head
    
    linkedList.deleteNode(0)
    print(linkedList)
    
    # Deletion from mid
    
    linkedList.deleteNode(5)
    print(linkedList)
    
    # deletion from last
    linkedList.deleteNode(9)
    print(linkedList)

File: Linked_List/Singly Linked List/LinkedList.java
import java.util.*;


class Linked_List{

    private static Scanner sc = new Scanner(System.in);
    
    public static void main(String[] args) {
        
        MyLinkedList linkedlist = new MyLinkedList();
        for(int i = 0;i<10;i++){
            linkedlist.insertAtLast(i);
        }

        System.out.println(linkedlist);

        // Deletion From Head
        linkedlist.deleteNode(0);
        System.out.println(linkedlist);

        // Deletion from mid

        linkedlist.deleteNode(5);
        System.out.println(linkedlist);

        // Deletion from last
        linkedlist.deleteNode(9);
        System.out.println(linkedlist);

        sc.close();
    }
}

class Node{

    int data;
    Node next;

    Node(int data){
        this.data = data;
        this.next = null;
    }
}

class MyLinkedList{

    Node head;

    MyLinkedList(){
        this.head = null;
    }
    

    public void insertAtLast(int data){
        Node insertionNode = new Node(data);

        // if our head is null 
        // No insertion yet
        if(this.head == null){
            // Our current node will be head
            this.head = insertionNode;
        }
        else{
            Node headPointer = this.head;

            while(headPointer.next != null){
                headPointer = headPointer.next;
            }

            headPointer.next = insertionNode;
        }
    }

    public String toString(){
        String result = "";

        Node headPointer = this.head;

        while(headPointer != null){
            if(headPointer.next != null){
                result += headPointer.data + " -> ";
            }
            else{
                result += headPointer.data;
            }
        }

        return result;
    }

    public void deleteNode(int dataKey){
        Node prevNode = null;
        Node currNode = this.head;
        
        // Deletion from head 
        if(currNode.data == dataKey){
            currNode = currNode.next;
            this.head = currNode;
            return;
        }
        // Deletion from other positions
        while(currNode!= null){
            if(currNode.data == dataKey){
                prevNode.next = currNode.next;
                return;
            }
            prevNode = currNode;
            currNode = currNode.next;
        }

        System.out.println("Node with this data not found");
    }

}

// 